conductor/lib/uploader.py

```diff
@@ -261,28 +325,43 @@ def do_work(self, job, thread_int):
             raise
 
     @common.DecRetry(retry_exceptions=api_client.CONNECTION_EXCEPTIONS, tries=5)
-    def do_upload(self, upload_url, filename, md5):
+    def do_singlepart_upload(self, upload_url, filename, file_size, md5):
         '''
         Note that for GCS we don't rely on the make_request's own retry mechanism because
         we need to recreate the chunked_reader generator before retrying the request.
         Instead, we wrap this method in a retry decorator.
 
-        We cannot reuse make_request method for S3 because it adds auth headers that
-        S3 does not accept. S3 also requires additional signatures to support chunked data.
-        For now S3 uploads will read all data in memory resulting in a smaller supported size.
+        We cannot reuse make_request method for S3 because it adds auth and Transfer-Encoding headers that
+        S3 does not accept.
         '''
 
         if "amazonaws" in upload_url:
-            headers = {'Content-Type': 'application/octet-stream'}
+            headers = {
+                'Content-Type': 'application/octet-stream',
+                'Content-Length': str(file_size),
```

Comment:
I could be wrong, but I believe the `requests` library automatically adds the content-length header for you in this circumstance.  In fact, if you add the the header yourself,  I think it doesn't matter what value you put in there; the `requests` lib will overwrite it with whatever it reads from the file (the body size). see https://github.com/psf/requests/blob/master/requests/models.py#L529

Reply:
ugh, let's leave a brief comment , e.g. # e.g. `must declare content-length ourselves due to zero byte bug in requests lib. see <wherever you left comment about bug>`